src/yarp.c

```diff
@@ -11218,42 +11235,109 @@ parse_expression_prefix(yp_parser_t *parser, yp_binding_power_t binding_power) {
+            yp_token_t opening_token = parser->previous;
 
-            if (quote == YP_HEREDOC_QUOTE_BACKTICK) {
-                assert(YP_NODE_TYPE_P(node, YP_INTERPOLATED_X_STRING_NODE));
-                yp_interpolated_xstring_node_closing_set(((yp_interpolated_x_string_node_t *) node), &parser->previous);
-                node->location = ((yp_interpolated_x_string_node_t *) node)->opening_loc;
-            } else {
-                assert(YP_NODE_TYPE_P(node, YP_INTERPOLATED_STRING_NODE));
-                yp_interpolated_string_node_closing_set((yp_interpolated_string_node_t *) node, &parser->previous);
-                node->location = ((yp_interpolated_string_node_t *) node)->opening_loc;
+            yp_node_t *part = parse_string_part(parser);
+
+            if (part == NULL) {
+                // We couldn't parse anything, so return a missing node
+                return (yp_node_t *) yp_missing_node_create(parser, parser->previous.start, parser->previous.end);
             }
 
-            // If this is a heredoc that is indented with a ~, then we need to dedent
-            // each line by the common leading whitespace.
-            if (indent == YP_HEREDOC_INDENT_TILDE) {
-                parse_heredoc_dedent(parser, node, quote);
+            if (YP_NODE_TYPE_P(part, YP_STRING_NODE) && match_any_type_p(parser, 2, YP_TOKEN_HEREDOC_END, YP_TOKEN_EOF)) {
+                // We only have a single string, so we can return it
+                yp_string_node_t *str_part = (yp_string_node_t *)part;
+                (str_part)->opening_loc = (yp_location_t) {
+                    .start = opening_token.start,
+                    .end = opening_token.end
+                };
+                (str_part)->closing_loc = (yp_location_t) {
+                    .start = parser->current.start,
+                    .end = parser->current.end
+                };
+                (str_part)->base.location = (yp_location_t) {
+                    .start = opening_token.start,
+                    .end = opening_token.end,
+                };
+                if (quote == YP_HEREDOC_QUOTE_BACKTICK) {
+                    part->type = YP_X_STRING_NODE;
+                }
+                lex_state_set(parser, YP_LEX_STATE_END);
+                expect(parser, YP_TOKEN_HEREDOC_END, YP_ERR_HEREDOC_TERM);
+                node = part;
+                if (indent == YP_HEREDOC_INDENT_TILDE) {
+                    int common_whitespace = parse_heredoc_common_whitespace_for_single_node(parser, node, -1);
+                    parse_heredoc_dedent_single_node(parser, &str_part->unescaped, true, common_whitespace, quote);
+                }
+            }
+            else {
+                // We have multiple parts, continue parsing them
+                yp_node_list_t parts = YP_EMPTY_NODE_LIST;
+
+                while (!match_any_type_p(parser, 2, YP_TOKEN_HEREDOC_END, YP_TOKEN_EOF)) {
+                    if (part != NULL) {
```

Comment:
The flow of this is kind of hard to read, because it looks like you're delaying appending parts onto the list based on the previous loop iteration. I think this might be easier if you append them right when you parse them.

```suggestion
                    if ((part = parse_string_part(parser)) != NULL) yp_node_list_append(&parts, part);
```

assuming before the loop that you `yp_node_list_append(&parts, part);`.